design/sales.md

```diff
@@ -45,3 +45,49 @@ the Reservations module, so that it survives node restarts.
     Datastore API:
       function put(key, value)
       function get(key): value
+
+Reserving storage space
+-----------------------
+
+When a request for storage is submitted on chain, the sales module decides
+whether or not it wants to act on it. If the requested duration and size match
+what the node wants to sell and the reward is sufficient, then the sales module
+will go through several steps to try and fill a slot in the request.
+
+First, it will select a slot from the request to fill. Then, it will reserve
+space in the Repo to ensure that it keeps enough space available to store the
+content. It will mark the reserved space as belonging to the slot. Next, it will
+download the content, calculate a storage proof, and submit the proof on chain.
+If any of these later steps fail, then the node should release the storage that
+it reserved earlier.
```

Comment:
I realize that this doesn't really fundamentally change what is being described here, but it is probably better to reserve a chunk of storage from the `RepoStore` and then slice it across several availabilities with different pricing strategies and durations. You almost never want to reserve right before filling a slot, since it most likely will make the process more complicated to handle at the code level, due to potential synchronization issues and other bookkeeping woes.

Reply:
> you can reserve a large chunk and then release in smaller chunks

In https://github.com/status-im/nim-codex/pull/340, we've been discussing a "new" model of Availabilities, such that a host will define how size, duration, and min price as CLI params at startup. At startup, we could reserve `size` bytes in the repo. Then once a request comes in, is matched to this availability, goes through its lifecycle until the end, what is meant to happen to the space that was occupied by that request? Are you saying to release it? If so, it would then be available for other node functions (eg caching). Also, how do we prevent duplication of reserved space on node restart?

Reply:
I think it makes most sense as described in the PR above for the CLI params to dictate a "maximum desired size for sale", which the reservations module will keep track of. Once a request comes in, only match that request (and reserve the space) if the size required for the slot storage plus the total space already reserved for the sales module would not exceed the "maximum desired size for sale". Once the request has completed its lifecycle, that space would be released.

Reply:
The repo has a concept of reserved and used, this where specifically added so that the sales module can "reserve" space that it wants to then offer for rent. Any amount can be reserved and released back to the repo at will. 

Let define what reserving vs renting means:

- Reserving: the repo allows reserving chunks of storage. This can be any size, but you probably want to reserve a large chunk and later release in smaller chunks (most likely block sized)
- Renting: availabilities allow further logically chunking the reserved space into configurable pieces that can be rented out under different conditions. For example, different duration vs size vs space parameters. 

The flow should look something like:

- Reserve some (large) chunk of space from the repo (this won't be usable by anything else)
- Slice the reserved chunk into several (or one) availabilities 
- When requests are matched against an availability, the host downloads data from the remote and writes it on drive
  - The host releases storage back to the repo in chunk/block sizes, right before writing data on drive. Releasing in small chunks avoids all sorts of concurrency issues that might arise (if this isn't enough, we can add atomic operations at the repo level, but this ins't something to worry about right now)

The above allows for a very simple an flexible mode of operation, where there is no need to do much at the repo level, just "reserve enough storage and release it on demand", and all the sales module needs to worry about is releasing back to the repo as much as it needs to write right before doing so.

> In https://github.com/status-im/nim-codex/pull/340, we've been discussing a "new" model of Availabilities, such that a host will define size, duration, and min price as CLI params at startup.

There is no reason to limit the system to the CLI, this should be dynamic parameters that can be tweaked at runtime. 

There are two separate configurable parameters:

- Total reserved space and
- The availabilities that map to that reserved space

Limiting these two options to the CLI, limits the ability to adjust this parameters without restarting the node. For example, a user might want to adjust the total number of reserved bytes (this should obviously not affect existing availabilities); or a user might want to change, remove or add new availabilities. Limiting this to CLI parameters doesn't really make much sense.

Reply:
Ok, I think i have a better understand of this, but please correct me if I'm wrong: storage bytes that have been written to disk need not be reserved, as they already count against the RepoStore's quota. That is why we can release chunks just before they are written to disk.

I had assumed incorrectly that bytes written to disk could still be reserved. 😅

> Renting: availabilities allow further logically chunking the reserved space into configurable pieces that can be rented out under different conditions. For example, different duration vs size vs space parameters.

Ok, you have a good point and now I understand why availabilities was always meant to be more than a single size, duration, min price.

As described in https://github.com/status-im/nim-codex/pull/340#issuecomment-1476448088, we should probably include the following in the spec as it describes quite well how availabilities would be treated after matching:
> I would think of availabilities strictly as pre-contract "offers for sale", in other words, this is local information that can be altered at any point by the storage node without affecting any global state and they get discarded and/or adjusted once a slot has been occupied. Discarded if an availability gets matched to a request 1:1, altered if a request uses part of an availability, in which case the available storage in the availability gets adjusted to reflect the matched request.

This brings up another point in that a host should be able to remove (or modify) availabilities that it no longer needs. An example would be if an availability offers 1000MiB for sale and a request uses 999MiB, the availability would get altered to be 1MiB in size, but with the same duration and size. The host may not think the duration and price are applicable and thus may want to delete (or modify) that availability.

> There is no reason to limit the system to the CLI, this should be dynamic parameters that can be tweaked at runtime.

I was always thinking of the CLI params as a starting point to reserve space in the repo, with the ability to modify at runtime if needed. You mentioned "the repo allows reserving chunks of storage. This can be any size, but you probably want to reserve a large chunk". How large of a chunk should be reserved if there is no cli param specifying this?

Reply:
> Ok, I think i have a better understand of this, but please correct me if I'm wrong: storage bytes that have been written to disk need not be reserved, as they already count against the RepoStore's quota. That is why we can release chunks just before they are written to disk.

Yep, thats correct - more info here https://github.com/status-im/codex-research/pull/161#pullrequestreview-1349654400

> As described in https://github.com/status-im/nim-codex/pull/340#issuecomment-1476448088, we should probably include the following in the spec as it describes quite well how availabilities would be treated after matching:

Yep, agree, it's probably good to include for clarity.

> I was always thinking of the CLI params as a starting point to reserve space in the repo, with the ability to modify at runtime if needed. You mentioned "the repo allows reserving chunks of storage. This can be any size, but you probably want to reserve a large chunk". How large of a chunk should be reserved if there is no cli param specifying this?

We already have a param for the total storage quota `storage-quota`, but we should probably add one for reservations as well.  There should be some sounds defaults, like a `10%` of the user's drive gets allotted to Codex, but the reserved quota should be only enabled if the salles/persistence options are set. I would still keep this simple, the majority of the admin of this should be exposed over REST APIs and the CLI should just enable some basic startup settings and defaults.

Reply:
> This brings up another point in that a host should be able to remove (or modify) availabilities that it no longer needs. An example would be if an availability offers 1000MiB for sale and a request uses 999MiB, the availability would get altered to be 1MiB in size, but with the same duration and size. The host may not think the duration and price are applicable and thus may want to delete (or modify) that availability.

Yep, good point.